src/compiler/scala/tools/nsc/typechecker/Infer.scala

```diff
@@ -589,8 +588,18 @@ trait Infer extends Checkable {
         // ...or lower bound of a type param, since they're asking for it.
         var checked, warning = false
         def checkForAny(): Unit = {
-          val collector = new ContainsAnyCollector(topTypes)
-          @`inline` def containsAny(t: Type) = t.dealiasWidenChain.exists(collector.collect)
+          val collector = new ContainsAnyCollector(topTypes) {
+            val seen = mutable.Set.empty[Type]
+            override def apply(t: Type): Unit = {
+              def saw(dw: Type): Unit =
+                if (!result && !seen(dw)) {
+                  seen += dw
+                  if (!dw.typeSymbol.isRefinementClass) super.apply(dw)
+                }
+              if (!result && !seen(t)) t.dealiasWidenChain.foreach(saw)
+            }
+          }
+          @`inline` def containsAny(t: Type) = collector.collect(t)
```

Comment:
The PR is green, with the same check files change as reverse(#8586), so LGTM.

There's a small change from reverse(#8586) (`collector.collect(t)` instead of `t.dealiasWidenChain.exists(collector.collect)`).  But looking at the impl, it looks like it was a redudant `dealiasWidenChain` anyways.

It helps that the PR is green and the check files diff is the same. 😄